### DISCUSSION TITLE
Use new branch drawing unicode characters for the graph?

### DISCUSSION THREAD

**Discussion by rbong** [SuperBo/fugit2.nvim#88] [Ideas]
There are new [branch drawing characters that were recently added to Kitty](https://github.com/kovidgoyal/kitty/pull/7681). Hopefully other terminals will follow suit and add these characters, and fonts will add backup glyphs as well.

Would you consider using these characters? If not, is there anything you'd change? It's still fairly experimental and we can get away with adding missing glyphs.

**Comment by SuperBo** [comment]
@rbong, Hello, I'm very glad to see you here :D. I think these symbols are more good-looking than ASCII box drawing symbols.

I still have some question, are these symbols work only on kitty? Will these characters be included in Nerd fonts?

**Reply by rbong** [reply]
Happy to be here... you've made some decisions with this plugin that I've always thought were the "proper" way to do a Git client in Vim.

Yes, these symbols work only in Kitty. Whether or not they get added to other terminals or fonts will depend on those projects... although it being in Kitty will be a strong point in its favor for getting it merged. There is also [flog-symbols](https://github.com/rbong/flog-symbols) which can be used for patching fonts.

If you'd like me to work on support before adding this, let me know. But it will also be useful if you can spot any issues with the character set now.

**Reply by SuperBo** [reply]
I have some unit tests to test git drawing, I will test these new symbols ASAP.

Can you share with me symbol table for git drawing characters. A symbol table like this [https://en.wikipedia.org/wiki/Box_Drawing](https://en.wikipedia.org/wiki/Box_Drawing). So that I can copy each symbol into code and test file.

**Reply by rbong** [reply]
Thanks, I added a [symbol table in flog-symbols](https://github.com/rbong/flog-symbols?tab=readme-ov-file#symbol-table)

**Comment by SuperBo** [comment]
Hi @rbong, these are first look of new unicode when applied to Fugit2 Graph.

<img width="884" alt="Screenshot 2024-08-03 at 14 04 45" src="https://github.com/user-attachments/assets/e340a395-2413-46b9-870b-4945c2323f0f">
<img width="884" alt="Screenshot 2024-08-03 at 14 04 09" src="https://github.com/user-attachments/assets/9edfe7df-1fe1-4e3b-a247-00fc91a35693">
<img width="884" alt="Screenshot 2024-08-03 at 14 04 03" src="https://github.com/user-attachments/assets/552c8ae6-f2ef-4de6-b70b-12d2d077d80e">
<img width="898" alt="Screenshot 2024-08-03 at 14 01 47" src="https://github.com/user-attachments/assets/a174066e-4725-4016-81ae-f89f7fe32df6"><img width="857" alt="Screenshot 2024-08-03 at 14 01 35" src="https://github.com/user-attachments/assets/3d188a76-a5c7-4961-ab2e-c539a314ebe9">

Here are some of my feedback:

- Can we make the vertical lines ![](https://github.com/rbong/flog-symbols/raw/main/img/uf5d1.png) solid (no space between lines).
- For commit symbols, can we make it larger, bolder?
- For commit symbols, I need a commit symbol with vertical branch and horizontal at the same time. Aka, can we combine these two symbols into one ![image0](https://github.com/rbong/flog-symbols/raw/main/img/uf5fa.png) + ![image1](https://github.com/user-attachments/assets/613db321-3f66-447e-aec5-50a614bb3cf2)
.

**Comment by rbong** [comment]
>Can we make the vertical lines solid (no space between lines).

Are you using the TTF font for patching? In Kitty it will line up perfectly. Otherwise, [according to the author of Kitty, it's impossible to get them to line up perfectly with all the different font options](https://github.com/kovidgoyal/kitty/discussions/7680#discussioncomment-10185245). This is why the feature was added to Kitty to draw these symbols itself. Sorry I forgot to mention this earlier, I hope you're still interested in the symbols.

I used the dimensions of box drawing characters in Nerd Fonts. Ie, if your box drawing characters will line up using pure Nerd Fonts, these symbols will line up. But I have found that some terminals draw box drawing characters themselves to get around this.

I tried a lot of combinations, but I could not find dimensions that worked in every font size on every terminal. If anyone can find more information on this I'm happy to change the dimensions of the symbols to something better.

>For commit symbols, can we make it larger, bolder?

I will make them bigger.

>For commit symbols, I need a commit symbol with vertical branch and horizontal at the same time

So it seems like we need all of these symbols, correct?

```
           |    | 
-o    o-  -o    o-
 |    |           

           |    | 
-*    *-  -*    *-
 |    |           

      |    |    | 
-*-  -*-  -*    *-
 |         |    | 

 | 
-*-
 | 

```

Where `o` is a normal commit and `*` is a merge commit.

---

I have a couple notes about symbol usage as well:

![2024-08-03-120324_108x78_scrot](https://github.com/user-attachments/assets/81abba6f-89b3-44df-adf4-7394b7d2ccba)

Concerning `U+F5DF` ( ![image](https://github.com/user-attachments/assets/2f0bd3eb-fed6-451d-9ec4-5e7f2268f337) ):

The intended symbol for this is `U+F5DE` ( ![image](https://github.com/user-attachments/assets/4ea97a2d-e060-4583-b522-954616aa8853) ). You are free to use the symbols however you want, but be aware this could look like the commit above and to the right isn't connected to the commit below it.

`U+F5DF` is intended for horizontally-aligned left-to-right graphs showing the above and below branches merging into the commit on the left.

![2024-08-03-113929_63x69_scrot](https://github.com/user-attachments/assets/2b08e5b8-841e-4392-a3ae-ff6b0266ed1b)

The intended usage for `U+F5D4` ( ![image](https://github.com/user-attachments/assets/0dd56034-255c-4813-aeae-ebf592f636f3) ) is for a branch "fading out" of the graph. It is intentionally not connected on the bottom, and I think we can make this look a lot better.

It seems like we could use symbols for branch crossovers. I know some branch viewers show one branch "hopping" over the horizontal line. Would you be happy with a character like this, or do you want the "dotted line" character?

Example of the "hopping" I'm talking about:

![image](https://github.com/user-attachments/assets/ff69e3c7-2419-4343-a3d3-2d626ce675f4)

**Reply by SuperBo** [reply]
- I'm using kitty (0.35.2). Do I need to update to nightly version to test it?
- For commit symbol, I don't think I will use all of combinations for now. But you can put all of them there for future usage (hope we still have enough of unicode space to put extra symbols).
- I agree with you on the usage of `U+F5DF` and `U+F5DE`, I will change it in my test code.
- I think I can use `U+F5D4` for lazy loading symbol :D, to show that user can load more if they want. Thank for your clarification.
- For `crossover` thing, I think a "hopping" symbol can cause distraction for user and it is not aesthetic in my opinion (hope you agree with me on this :D). I think a ┆ or ┅ will look nicer.

**Reply by rbong** [reply]
>I'm using kitty (0.35.2). Do I need to update to nightly version to test it?

Yes, there hasn't been a release since it was merged so you have to use nightly or build from source.

>I think a "hopping" symbol can cause distraction for user and it is not aesthetic in my opinion (hope you agree with me on this :D). I think a ┆ or ┅ will look nicer.

Honestly, I think that ┆ and ┅ don't communicate clearly to the user what is happening with the branch. I used to use them in Flog because I didn't have anything better, but as of right now I just do this, where the branch just doesn't touch the merge line:

![2024-08-03-143200_192x343_scrot](https://github.com/user-attachments/assets/32094869-fdd8-4213-865a-5046de312758)

From branch viewers not limited to the terminal, I have seen the "hopping" style or the "no touching" style. I have never seen anyone do the "dotted line" style in a GUI outside of the terminal.

You have 2 different options if we don't add this symbol:

1) In Kitty, you can always use the dotted line box drawing character itself (┆) and it will line up perfectly since it generates it using the same line drawing code. The only time it won't is if someone builds from source and does something funky.

2) If we add "branch crossover" characters to the TTF file, people could also always customize them.

Neither of these options are perfect. The first option potentially will cause the symbols to not line up in terminals other than Kitty (if they use the backup font symbols and get them to match up, that doesn't mean they'll be aligned with their terminal/font's box drawing characters). The second option isn't available in Kitty (and some other terminals), because Kitty will override these symbols so that they line up.

The third option is to add multiple styles of branch crossover symbols to the symbol set and include this "dotted" style. There is precedent for this - Powerline includes all kinds of different styles with different symbols, after all. However, this isn't Powerline. We have more than left/right characters. Providing different "styles" via separate symbols opens a can of worms - if we provide different "styles" for other things, the number of symbols could really get out of hand. Maybe this is a one time thing though.

I am not sure what option is best, let me know your thoughts after hearing my perspective.

**Reply by SuperBo** [reply]
![](https://private-user-images.githubusercontent.com/5875706/354836662-32094869-fdd8-4213-865a-5046de312758.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI3NDAzNzksIm5iZiI6MTcyMjc0MDA3OSwicGF0aCI6Ii81ODc1NzA2LzM1NDgzNjY2Mi0zMjA5NDg2OS1mZGQ4LTQyMTMtODY1YS01MDQ2ZGUzMTI3NTgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgwNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MDRUMDI1NDM5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YWRlNWM5NzlmZmRiMjI3MTc4YjE3MDc4YWU3OTZiODQzYTIxNDQ5MDY3MDhmYzJmNjEwZTgwYmI4ZWNlNTRlOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.ep5jdfNPvyHJJNyDOcc8oHX7LJwV11g11hLCOzwMIQo) this one look nice. I will use this instead of dotted line 🥹. You an test other crossover symbol (hopping, dotted), however, I'm happy with simple line method.

**Comment by SuperBo** [comment]
Good morning @rbong, can I ask for some extra symbol? I'm working with Git Rebase graph, in which you can show git graph beside rebase action. Could you support fixup commit and squash commit symbol :D

![Screenshot 2024-08-05 at 07 16 51](https://github.com/user-attachments/assets/bd0b1081-e25a-40d7-93ec-b4e96157701e)

**Reply by rbong** [reply]
Hey SuperBo, question about this view. Is it ever possible to show other commits or anything other than a straight line in this view? I'm asking because the Git CLI doesn't have a strict equivalent for this view, it can just show a list of rebase actions without the graph.

**Reply by SuperBo** [reply]
I think straight line is enough. Git CLI only shows ordered list of commits and hence implies a straight line of graph.

**Reply by rbong** [reply]
Thanks, I just wanted to figure out how this view is supposed to work.

I personally think it might not make sense to have rebase symbols as part of the symbol set if it's always going to be a straight line. The idea behind the symbols is to make it possible to draw branch direction in Git commit graphs. But if it's always just a straight line, there's never going to be a complex graph to draw.

In the interests of not having the symbol set blow up in size, I personally think we should not add these symbols. I'd like to get other people's opinions though.

**Comment by rbong** [comment]
I added the new symbols we discussed to [flog-symbols](https://github.com/rbong/flog-symbols) and created a PR to Kitty with the new symbols and larger commit circles.
Sorry it took so long. Let me know if you have any feedback.

**Reply by SuperBo** [reply]
sorry @rbong, I've just come back to github.

**Reply by SuperBo** [reply]
Does this flog-symbols become mainstream and supported in nerdfont and kitty?

**Reply by rbong** [reply]
Not supported in nerdfont yet, but it's in some terminals, including Kitty. Maybe it's about time to submit a PR to Nerdfont - it will have alignment issues in most terminals still but might help adoption and work in some cases.
